ValidateSetAttribute enhancement: support set values to be dynamically generated from a custom ValidateSetValueGenerator

iSazonov opened the pull request:
Close #3744 

Currently `ValidateSetAttribute` accepts only explicit constants as valid values. This is a significant limitation. Sometimes we need to get valid values dynamically, ex., Azure VMs, logged-on users and so on. The PR add follow possibilities:
- pass a _custom type_ (a valid values generator) implementing `IValidateSetValuesGenerator` interface to get valid values dynamically.
- pass a _custom type_ (a valid values generator) derived  from `CachedValidValuesGeneratorBase` abstract class to get valid values dynamically and _cache_ the list to share with other ValidateSetAttribute attributes.

We support valid values generators on PowerShell and C#.

iSazonov committed:
Add dynamically generated set in ValidateSetAttribute for binary cmdlets

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1456,43 @@ public ValidateSetAttribute(params string[] validValues)
+        public ValidateSetAttribute(Type type)
+        {
+            IValidateSetValuesGenerator validValuesGenerator;
+            if (typeof(IValidateSetValuesGenerator).IsAssignableFrom(type))
```
You can use the `as` operator.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1456,43 @@ public ValidateSetAttribute(params string[] validValues)
+            }
+            else
+            {
+                throw PSTraceSource.NewArgumentNullException("validValues");
```
Cut and paste? I would use `ArgumentException` and please use the correct parameter name.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1456,43 @@ public ValidateSetAttribute(params string[] validValues)
+                throw PSTraceSource.NewArgumentNullException("validValues");
+            }
+
+            _validValues = validValuesGenerator.GetValidValues().ToArray();
```
We shouldn't cache the values during construction - the set may change over time.

Also - we may want a way to query other bound parameters from the generator - which would not be possible in the constructor.

iSazonov replied:
I eliminated the caching.

Please clarify - do you say to put `validValuesGenerator` in a property?

lzybkr commented on test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1:
```diff
@@ -216,3 +218,96 @@ Describe 'Type resolution with attributes' -Tag "CI" {
+                    }
+                }
+
+                /// Implement of test IValidateSetValuesGenerator to return Null
```
This implementation does not return `null` - it returns an empty list.

iSazonov replied:
Could you help me here?

lzybkr replied:
I recommend learning more about how the `yield` keyword is implemented under the covers, e.g. this is a good article: http://csharpindepth.com/Articles/Chapter6/IteratorBlockImplementation.aspx

iSazonov replied:
Thanks for useful link!

Test was removed.

lzybkr commented on test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1:
```diff
@@ -216,3 +218,96 @@ Describe 'Type resolution with attributes' -Tag "CI" {
 }
+
+Describe 'ValidateSet support a dynamically generated set' -Tag "CI" {
+    Context 'C# test' {
```
You should also provide a test implementing the interface in PowerShell.

iSazonov replied:
Do you have in mind something like?
```powershell
            function Get-TestValidateSet3
            {
                [CmdletBinding()]
                Param
                (
                    [ValidateSet([type]::GetType("GenValuesForParam1"))]
                    $Param1
                )

                $Param1
            }
```

lzybkr replied:
No - I meant:
```powershell
class GenValuesFromParam2 : IValidateSetValuesGenerator { ... }
...
[ValidateSet([GenValuesFromParam2])] ...
```
In other words - no Add-Type.

iSazonov replied:
Following sample don't work - the type passed as string to `public ValidateSetAttribute(params string[] validValues)` constructor - wrong binding 😕 :
```powershell
            class GenValuesForParam2 : System.Management.Automation.IValidateSetValuesGenerator {
                [System.Collections.Generic.IEnumerable[String]] GetValidValues() { return "Test1","TestString","Test2" }
            }

            function Get-TestValidateSet3
            {
                [CmdletBinding()]
                Param
                (
                    [ValidateSet([GenValuesForParam2])]
                    $Param1
                )

                $Param1
            }
```

iSazonov replied:
Fixed.

iSazonov committed:
Remove caching valid values and unneeded test

iSazonov committed:
Add "powershell" test

iSazonov committed:
Dynamic binding for ValidateSetAttribute

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1453,39 @@ public ValidateSetAttribute(params string[] validValues)
+                throw PSTraceSource.NewArgumentException("valuesGeneratorType");
+            }
+
+            _validValues = validValuesGenerator.GetValidValues();
```
The constructor should just save `valuesGeneratorType` because we create just one instance of the attribute per use.

For example, if someone added `ValidateSet` to the `Name` parameter on `Get-VM` - you would want the name of each vm whenever you called `Get-VM`, not the names when you first called `Get-VM`.

By constructing the instance and getting the values during validation, the set is dynamically generated when tab completion or validation is happening - which only aligns with construction once (for scripts, attributes are constructed at first invocation) or never (for binary cmdlets, attributes are eagerly constructed).

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1359,7 +1360,7 @@ public ValidateCountAttribute(int minLength, int maxLength)
     [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
     public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
     {
-        private string[] _validValues;
```
I think this shouldn't change, and I would add a new read-only property for the Type so you can tell the difference on which constructor was used.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1386,7 +1387,7 @@ public IList<string> ValidValues
             get
             {
-                return _validValues;
+                return _validValues.ToList();
```
For the dynamic case - this should probably return a new collection every time it's called.

lzybkr reviewed (changes requested):
One other thought - some of the dynamic sets are expensive to compute, e.g. querying Azure or whatever.

One strategy is to cache the set for some time. Nothing in this design prevents such caching, but it would be better to implement the caching in PowerShell so that third party code can use it easily.  Maybe that means adding a `CacheExpiration` parameter that defaults to 0 - immediate expiration.

Still thinking out loud - the cache maybe belongs in the generator, but maybe we provide some helpers to manage it - perhaps as an abstract base class.

iSazonov committed:
Add caching valid values and tests

iSazonov commented:
>Still thinking out loud - the cache maybe belongs in the generator, but maybe we provide some helpers to manage it - perhaps as an abstract base class.

Wouldn't that be too intrusive? `IValidateSetValuesGenerator` looks very powerful.
If we want such cache in the generator it seems we should migrate from `IValidateSetValuesGenerator` interface to `ValidateSetValuesGenerator` class and force users to inherite from the class. Maybe we stop on `IValidateSetValuesGenerator`? Now we have 'generic' caching internally in `ValidateSet` so maybe we'll leave the users full freedom in `IValidateSetValuesGenerator`?

lzybkr commented:
I had feedback in the form of a [PR to TabExpansionPlusPlus](https://github.com/lzybkr/TabExpansionPlusPlus/commit/b1d99d34aac92b07fa76aae3c5aa5a426e75993c)  that caching should be built-in.

iSazonov commented:
Do you mean that the cache has to work between different `ValidateSet` instances? With `validValuesGeneratorType` key?

lzybkr commented:
Many cmdlets have parameters that are common, e.g. there are 72 commands in AzureRM.Compute that accept `-ResourceGroupName`.

On my machine, it takes >1s to query the resource group names

So I think a cache that works across `ValidateSet` instances would be very helpful.

powercode commented:
@lzybkr 
> I had feedback in the form of a PR to TabExpansionPlusPlus that caching should be built-in.
It feels almost more often than not that I implement some sort of custom completion caching. But often with less than stellar aging/expiration.

iSazonov committed:
Add global cache for valid values

iSazonov commented:
I had to rebase to get autoloading of helper modules.

Global valid values cache was added. The only question is. I clean up this cache never. Should we add clean-up and where? Although it seems superfluous.

iSazonov commented:
@lzybkr Could you please continue the code review?

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```
Is there a good reason to store the values in the field instead of `return validValuesCacheEntry.validValues;`?

lzybkr replied:
I think it would also be useful to clean up the cache - this could be done with something like `Task.Delay(CacheExpiration).ContinueWith(() => validValuesCacheEntry.validValues = null);`

This might not seem like a big deal, the collection should typically be small, but sometimes it might be large, so we should make the collection collectible if it's truly going to be used again.

iSazonov replied:
>Is there a good reason to store the values in the field instead of  return validValuesCacheEntry.validValues; ?

We only use this in `SetAsString()` for error message. We could use `ValidValues` if we're not afraid that there will be a cache expiration and a recalculation of  `ValidValues` and that there may be other data in the error message.

>  Task.Delay(CacheExpiration)

Different attributes can have different `CacheExpiration`. How can we resolve the problem?

iSazonov replied:
Could you please continue?
It seems we should clean up validValuesCache from expired Entries not validValuesCacheEntry.validValues.

lzybkr replied:
Sorry, I thought I had replied.

I now think `CacheExpiration` should **not** be an argument to the attribute because, as you point out, different uses might specify conflicting values.  Instead, the generator should own the cache expiration policy, and the attribute should not cache the results from the generator.

I still like the idea of providing an abstract base class that implements a default caching policy - it would be optional for a generator to derive from it.

iSazonov replied:
Can we add this to  IValidateSetValuesGenerator  interface?

And please comment my "It seems we should clean up validValuesCache from expired Entries (generators) ".

lzybkr replied:
Sure, you could add `CacheExpiration` to `IValidateSetValuesGenerator`, but it would be something extra that some generators have no need for.

I agree we should clean up `validValuesCache` - and ideally, I think that cache is optionally available to generators but not required. This is why I suggested an abstract base class that generators could derive from.

iSazonov replied:
@lzybkr I still don't correct tests however please review my next step - Is that the right direction?

I removed valid values cache and  CacheExpiration  from ValidateSet attribute, enhance a generator cache (add cleanup - the idea is from [CoreFX](https://github.com/stephentoub/corefx/blob/9820567e1bb1972502bfe6345963eb0ab99ef3f9/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs)), add an abstract class for generators with valid values cache support.

iSazonov replied:
@lzybkr Now tests adjusted.

lzybkr reviewed (changes requested):
This is starting to look pretty good.

I have a feeling that this will make many folks happy, but it will fall short for others, mostly around wanting access to some context, e.g. the cmdlet or some of the already bound parameters.

Hopefully we can get some feedback one way or the other on the usefulness of this as it is though.

iSazonov committed:
Fix 'NewValidateSetAttribute' and add test for unimplemented generator type

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -739,7 +739,6 @@ static Compiler()
             s_builtinAttributeGenerator.Add(typeof(ParameterAttribute), NewParameterAttribute);
             s_builtinAttributeGenerator.Add(typeof(OutputTypeAttribute), NewOutputTypeAttribute);
             s_builtinAttributeGenerator.Add(typeof(AliasAttribute), NewAliasAttribute);
-            s_builtinAttributeGenerator.Add(typeof(ValidateSetAttribute), NewValidateSetAttribute);
```
I temporary removed that to pass tests but I forgot to ask you about that.
Now fixed.

lfshr commented:
Is this definitely limited to 'binary' cmdlets? Has anyone tested this using a PS declared class?

iSazonov commented:
@lfshr Scripts is supported - see tests in the PR. We will edit the PR description after completing this work.

lfshr commented:
@iSazonov absolutely brilliant, v good work! Well done @lzybkr for the solution. (I take full credit for the moan that instigated it 😛)

schittli commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  thomas.schittli@akros.ch
    Domain jig.ch has exceeded the max emails per hour (209/200 (104%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 5430 characters long; only the first
------ 5000 or so are included here.
Received: from o4.sgmail.github.com ([192.254.112.99]:33553)
	by rs111.nsresponse.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89)
	(envelope-from <bounces+848413-aac0-github=jig.ch@sgmail.github.com>)
	id 1dJ3YW-000ecy-9S
	for github@jig.ch; Thu, 08 Jun 2017 14:57:22 -0500
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=hShJWcvT9Pb6uxla7Fnd9iixHHM=; b=Y3gswrT8hSHg3Sug
	3nESvwqFy3i0f7kr1bXJ62TWl0X//POn4qUMScz6eJ9NDbdycbPyASBXc4zv8B7T
	nHUY0pDTfrXlDmwYHgDei3Y9FrXvzTi1dsgppciKhx2QZS0Nq19xTbjzEnkFH9Vc
	WoLLO7uWPKx1QBTyMC8ZholprWs=
Received: by filter0986p1mdw1.sendgrid.net with SMTP id filter0986p1mdw1-9547-5939ABF5-38
        2017-06-08 19:56:37.867097011 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0005p1iad1.sendgrid.net (SG) with ESMTP id exs5qSK6Sb6QZKnfIy_g2Q
	for <github@jig.ch>; Thu, 08 Jun 2017 19:56:37.840 +0000 (UTC)
Date: Thu, 08 Jun 2017 12:56:37 -0700
From: Jason Shirk <notifications@github.com>
Reply-To: PowerShell/PowerShell <reply@reply.github.com>
To: PowerShell/PowerShell <PowerShell@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <PowerShell/PowerShell/pull/3784/review/42996423@github.com>
In-Reply-To: <PowerShell/PowerShell/pull/3784@github.com>
References: <PowerShell/PowerShell/pull/3784@github.com>
Subject: Re: [PowerShell/PowerShell] Add dynamically generated set in
 ValidateSetAttribute for binary cmdlets (#3784)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5939abf5a8c90_eb33fcd63a97c3c955a8";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: lzybkr
X-GitHub-Recipient: schittli
X-GitHub-Reason: subscribed
List-ID: PowerShell/PowerShell <PowerShell.PowerShell.github.com>
List-Archive: https://github.com/PowerShell/PowerShell
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+007e62312cd24a72941b17737850ee1cc647e6117d3448aa92cf0000000115516df592a169ce0da0be8e@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AH5iMem0-KOQppPU-TNvlDnqKHpXiN22ks5sCFH1gaJpZM4Nazg4>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: github@jig.ch
X-SG-EID: DAOE8jsA1R6juA5Oy75tmkOSGcm/2gf22STf7FTBCw/DIaYvSgqsQIQbJ5dG6AlB/z5qcYlJTCG5FM
 dlENOGDddCcMVTWy32QMl98dlaCymYgMf5SsqKmZmWgA6ODAfB02J179/rg5geiDt0CpQKkz1Z/N3Q
 sOFFSENCfxSqD6sEkTCCsbvDwMaAWGnd4kdAY8hsTaznGuPsAEndVYfd6A==
X-Spam-Status: No, score=-3.0
X-Spam-Score: -29
X-Spam-Bar: ---
X-Ham-Report: Spam detection software, running on the system "rs111.nsresponse.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  lzybkr commented on this pull request. > + else + { + var
   validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
    + var validValues = validValuesGenerator.GetValidValues()?.ToArray(); + var
    cacheEntryCreationTime = DateTime.Now; + validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator,
    validValues, cacheEntryCreationTime); + validValuesCache.TryAdd(validValuesGeneratorType,
    validValuesCacheEntry); + } + if (validValuesCacheEntry.validValues == null)
    + { + throw new ValidationMetadataException( + "ValidateSetGeneratedValidValuesListIsEmpty",
    + null, + Metadata.ValidateSetGeneratedValidValuesListIsEmpty); + } + _validValues
    = validValuesCacheEntry.validValues; [...] 
 
 Content analysis details:   (-3.0 points, 5.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
  0.0 URIBL_BLOCKED          ADMINISTRATOR NOTICE: The query to URIBL was blocked.
                             See
                             http://wiki.apache.org/spamassassin/DnsBlocklists#dnsbl-block
                              for more information.
                             [URIs: github.com]
 -0.0 SPF_PASS               SPF: sender matches SPF record
 -0.0 T_RP_MATCHES_RCVD      Envelope sender domain matches handover relay
                             domain
 -1.0 RCVD_IN_MSPIKE_H2      RBL: Average reputation (+2)
                             [192.254.112.99 listed in wl.mailspike.net]
 -1.9 BAYES_00               BODY: Bayes spam probability is 0 to 1%
                             [score: 0.0000]
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
X-Spam-Flag: NO

----==_mimepart_5939abf5a8c90_eb33fcd63a97c3c955a8
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

lzybkr commented on this pull request.



> +                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;

Sure, you could add `CacheExpiration` to `IValidateSetValuesGenerator`, but it would be something extra that some generators have no need for.

I agree we should clean up `validValuesCache` - and ideally, I think that cache is optionally available to generators but not required. This is why I suggested an abstract base class that generators could derive from.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/PowerShell/PowerShell/pull/3784#discussion_r120983547
----==_mimepart_5939abf5a8c90_eb33fcd63a97c3c955a8
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><b>@lzybkr</b> commented on this pull request.</p>

<hr>

<p>In <a href="https://github.com/PowerShell/PowerShell/pull/3784#discussion_r120983547">src/System.Management.Automation/engine/Attributes.cs</a>:</p>
<pre style='color:#555'>&gt; +                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            &quot;ValidateSetGeneratedValidValuesListIsEmpty&quot;,
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
</pre>
<p>Sure, you could add <code>CacheExpiration</code> to <code>IValidateSetValuesGenerator</code>, but it would be something extra that some generators have no need for.</p>
<p>I agree we should clean up <code>validValuesCache</code> - and ideally, I think that cache is optionally available to generators but not required. This is why I suggested an abstract base class that generators could derive from.</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/PowerShell/PowerShell/pull/3784#discussion_r120983547">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AH5iMRCqNFwj-c8feAxOZKIHYdG7MOfwks5sCFH1gaJpZM4Nazg4">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AH5iMYyRloZK_Ku0SGeDYZrrbOa1T3jdks5sCFH1gaJpZM4Nazg4.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/PowerShell/PowerShell/pull/3784#discussion_r120983547"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

<script type="application/json" data-scope="inboxmarkup">{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/PowerShell/PowerShell","title":"PowerShell/PowerShell","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.github

iSazonov committed:
Remove a valid values cache from ValidateSet attribute

1. Remove a valid values cache from ValidateSet attribute.
2. Enhance a valid value generator cache.
3. Add an abstract class for valid value generator with valid values
cache support.

iSazonov committed:
Refactor CachedValidValuesGeneratorBase class

Remove CacheExpiration tests
Add CachedValidValuesGeneratorBase class tests

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
+            if (ValidValuesCacheExpiration > 0)
+            {
+                Task.Delay(ValidValuesCacheExpiration).ContinueWith((task) => _validValues = null);
+                _validValues = _validValuesNoCache;
```
The parameter to `Task.Delay` is in milliseconds, not seconds.

lzybkr replied:
We could, but I think that seconds is a more practical unit for the timeout.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
+                return _validValues;
+            }
+
+            var _validValuesNoCache = GenerateValidValues();
```
>_validValuesNoCache [](start = 16, length = 19)

Local variables should not use a `_` prefix.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
+            if (_validValues != null)
+            {
+                return _validValues;
+            }
```
There is a race condition here.

The sequence to trigger the race is:

1. _validValues != null
2. The background task to release the cache runs, assigning _validValues with null.
3. The return executes, returning null.

The fix is:

```c#
var validValuesLocal = _validValues;
if (validValuesLocal != null)
{
    return validValuesLocal;
}
```

iSazonov replied:
Thanks! I thought about it, but I didn't know how to just fix it.

Fixed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
         private string[] _validValues;
 
+        // The valid values generator cache works across 'ValidateSetAttribute' instances.
+        private static ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry>();
```
>ValidValuesGeneratorCacheEntry [](start = 50, length = 30)

I think this can be `IValidateSetValuesGenerator`.
There will be very few types that implement this interface, so I think the extra complexity around cache cleanup on the generator is not needed.

lzybkr replied:
I'm a little confused. The is cache cleanup not controlled by the user - so restarting PowerShell is a practical necessity.

iSazonov replied:
We can use `ValidValuesGeneratorCacheExpiration` to implicitly remove expired values from the cache.
Main question is - will we be happy with the PowerShell restart?

lzybkr replied:
Restarting PowerShell is an occasional annoyance, but I'm not sure this will make a difference.

For C# implementations, you need to restart anyway.
For script implementations - I think there are acceptable ways to develop your generator and test it without restarting PowerShell.

So my preference is to keep the code as simple as possible and only introduce additional complexity after gaining more experience with the feature and determining it really is necessary.

iSazonov replied:
I agree that we can wait a feedback from users.
The cache is removed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1386,6 +1532,40 @@ public IList<string> ValidValues
+                        // We don't cache valid values.
+                        // We expect that valid values can be cached in the valid values generator.
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var cacheEntryAccessTime = DateTime.Now;
```
I think I'd move creating the generator and storing it in the ConcurrentDictionary to the ValidateSet constructor.

iSazonov replied:
Yes, now the cache is removed and we can create the generator in the constructor.
Fixed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1386,6 +1532,40 @@ public IList<string> ValidValues
+                    }
+
+                    _validValues = ValidValuesGeneratorCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+
```
As long as we're always call `ToArray`, maybe we should just change the interface to return `string[]`.

iSazonov replied:
Do we really want tochange _public_ interface? Is the breaking change?

lzybkr replied:
Sorry for not being clear - I meant the interface introduced in this PR - `IValidateSetValuesGenerator` - so not a breaking change.

iSazonov replied:
I agree - the more simple for scripts.

iSazonov committed:
Fix race condition, fix local name

iSazonov committed:
Fix validValuesGenerator
Change IValidateSetValuesGenerator (return type string{})
Now ValidValuesCacheExpiration in seconds
Remove validValuesGeneratorCache cleanup
Move generator creation to constructor
Fix tests

iSazonov commented:
@lzybkr Could you please continue the code review?

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Base class for all valid values generators with cache support.
```
I'd change this comment to something like:

```
Optional base class for <see cref="IValidateSetValuesGenerator"/> implementations that want a default implementation to cache values.
```

iSazonov replied:
Done.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
+            }
+        }
+
+        private int _ValidValuesCacheExpiration;
```
`_v`, and add a blank line.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,13 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
+        /// Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.
+        /// By default 'ValidValuesCacheExpiration = 0' and no valid values is cached (is re-evaluated every time it is used).
+        /// </summary>
+        public int ValidValuesCacheExpiration { 
```
I think this class should have a single constructor:
```C#
public CachedValidValuesGeneratorBase(int cacheExpirationInSeconds) { _validValuesCacheExpiration = cacheExpirationInSeconds; }
```

I don't think the property should be exposed - as changing the value implies something (e.g. that I can extend or shorten the cache expiration for values already cached.)

lzybkr replied:
Maybe we do need a method to clear the cache though - I can see some external event happening and the generator wanting to drop the cache.

iSazonov replied:
The property is not available for users from outside the class because we create an instance of the class by type with _default_ constructor only. We need way to set `_validValuesCacheExpiration`. See a sample from our tests below. So we cannot remove the property but we can make it as `protected`. 

```powershell
            class GenValuesWithExpiration : System.Management.Automation.CachedValidValuesGeneratorBase {
                GenValuesWithExpiration() : base() {
                    $this.ValidValuesCacheExpiration = 2
                }

                Static [bool] $temp = $true;

                [String[]] GenerateValidValues() {

                    if ([GenValuesWithExpiration]::temp) {
                        [GenValuesWithExpiration]::temp = $false
                        return [string[]]("Test1","TestString1","Test2")
                    } else {
                        [GenValuesWithExpiration]::temp = $true
                        return [string[]]("Test1","TestString2","Test2")
                    }
                    
                }
            }
```

lzybkr replied:
I'm suggesting:

```C#
    class GenValuesWithExpiration : System.Management.Automation.CachedValidValuesGeneratorBase {
        GenValuesWithExpiration() : base(2) {}
        # Rest is the same
```

The PowerShell engine is calling the default constructor, but the abstract base doesn't need a default constructor.

iSazonov replied:
Thanks for clarify! Is it good UX? For me, this is not as intuitive as using the property name explicitly.

lzybkr replied:
UX is bad if you provide an option to set the property but semantics of setting are fuzzy. I think it's better to not expose the setter in this case.

iSazonov replied:
Thanks!

The property is removed.

iSazonov replied:
`ClearValidValuesGeneratorCache` was added.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1542,39 @@ public ValidateSetAttribute(params string[] validValues)
+                throw PSTraceSource.NewArgumentException("valuesGeneratorType");
+            }
+
+            validValuesGeneratorType = valuesGeneratorType;
```
Why store the type? If we stored the instance, we only look up in `s_ValidValuesGeneratorCache` once per use of the attribute instead of once per invocation of the command using the attribute.

iSazonov replied:
Sorry, I don't understand the question.
We use `validValuesGeneratorType` in `ValidValues` property.

lzybkr replied:
You only need to store the type if you defer creating the instance. If you create and store the instance in the constructor, I think that's better overall.

iSazonov replied:
Thanks for clarify! I see now.
Won't there be a problem if we add a cache cleanup?

lzybkr replied:
I don't anticipate any problems - these objects are small assuming they leave caching to us.

iSazonov replied:
Done.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1298,15 +1298,35 @@ private static Attribute NewAliasAttribute(AttributeAst ast)
+                else
+                {
+                    throw InterpreterError.NewInterpreterException(ast, typeof(RuntimeException), ast.Extent,
+                        "CustomValidValuesGeneratorTypeNotFound", ParserStrings.CustomValidValuesGeneratorTypeNotFound, generatorTypeAst.TypeName.FullName, typeof(System.Management.Automation.IValidateSetValuesGenerator).FullName);
```
It's probably better to reuse `TypeNotFound` instead of adding a new error message.

iSazonov replied:
Done.

lzybkr replied:
If we didn't have this special case code, you would have gotten `TypeNotFound`, which is exactly the problem - it has nothing to do with `IValidateSetValuesGenerator`. Besides, if you don't implement the interface, that error is reported when constructing the attribute.

iSazonov replied:
Thanks! 

Reuse  TypeNotFound done.

iSazonov replied:
Sorry, VS Code removed trailing spaces.

lzybkr commented on src/System.Management.Automation/resources/Metadata.resx:
```diff
@@ -168,6 +168,9 @@
   <data name="ValidateSetFailure" xml:space="preserve">
     <value>The argument "{0}" does not belong to the set "{1}" specified by the ValidateSet attribute. Supply an argument that is in the set and then try the command again.</value>
   </data>
+  <data name="ValidateSetGeneratedValidValuesListIsEmpty" xml:space="preserve">
```
The ErrorId doesn't match the message - is it `null` or empty?

iSazonov replied:
Null.

Fixed.

lzybkr commented on src/System.Management.Automation/resources/ParserStrings.resx:
```diff
@@ -759,6 +759,9 @@ Possible matches are</value>
   <data name="CustomAttributeTypeNotFound" xml:space="preserve">
     <value>Cannot find the type for custom attribute '{0}'. Make sure that the assembly that contains this type is loaded.</value>
   </data>
+  <data name="CustomValidValuesGeneratorTypeNotFound" xml:space="preserve">
```
I think I'd reuse `TypeNotFound` instead of adding a new message.

iSazonov replied:
Removed.

lzybkr reviewed (dismissed):
Great - this is looking pretty good now, just a few little things to clean up I think.

iSazonov committed:
Minor corrections

1. Now 'ValidValuesCacheExpiration' is 'protected'
2. Correct resource string
3. Correct tests

iSazonov committed:
Reuse TypeNotFound

iSazonov committed:
Remove `validValuesGeneratorType`. Add `CachedValidValuesGeneratorBase` constructor

iSazonov committed:
Add 'ClearValidValuesGeneratorCache'

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,19 +1357,90 @@ public ValidateCountAttribute(int minLength, int maxLength)
+        /// <summary>
+        /// Clear the valid values generator cache.
+        /// </summary>
+        public static void ClearValidValuesGeneratorCache()
```
I'm not sure this is useful - because each instance of the attribute will also contain a reference to the `IValidateSetValuesGenerator` and we can't clear that.

When I suggested a method to clear the cache, I was referring to `CachedValidValuesGeneratorBase`, but I think it's probably fine to not add that - one can simply use a smaller timeout value or implement their own custom caching logic.

iSazonov replied:
Sorry, my attention is focused on the current virus attacks. 
I remember that we rejected the clear already.

Removed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,19 +1357,90 @@ public ValidateCountAttribute(int minLength, int maxLength)
+
+            if (_validValuesCacheExpiration > 0)
+            {
+                Task.Delay(_validValuesCacheExpiration * 1000).ContinueWith((task) => _validValues = null);
```
There is an extremely unlikely race condition here - the task could complete before the assignment on the next line. I can't see it ever happening except maybe on a single core CPU where PowerShell gets suspended (say machine goes to sleep) at just the right time, and when it resumes, the background task gets to run first.

So I'd swap these two lines - I don't see any downsides to doing that.

iSazonov replied:
I swaped the lines.

Please clarify. If you mean that the race condition is related with clearing `_validValues` - we add a protect above (validValuesLocal) and return validValuesNoCache below - so clearing `_validValues` should not cause problems.

lzybkr replied:
This would be a different race - we wouldn't actually clear the cache.

lzybkr reviewed (commented):
Thanks for addressing all of my feedback, I think this is almost ready.

iSazonov committed:
Remove `ClearValidValuesGeneratorCache`

iSazonov commented:
> I have a feeling that this will make many folks happy, but it will fall short for others, mostly around wanting access to some context, e.g. the cmdlet or some of the already bound parameters.

@lzybkr Shall we continue with the "some context"?

iSazonov commented:
It seems we catch a race condition on slow OSX. 😕

lzybkr commented:
Wasn't there a CLR bug we were hitting on Mac related to defining classes? Maybe @daxian-dbw remembers.

lzybkr commented:
If you want to investigate options for providing more information to the generator - I could imagine wanting to know the command, maybe the command ast, the actual parameter the attribute is applied to, and maybe the already bound parameters.

iSazonov commented:
I tried to think about it. If we cache and share the generator, we can't store the context in it and should pass it as a parameter.  Even if we send `engineIntrinsics` from `Validate()` to `ValidateElement` we cannot send it in a generator - we get valid values with the `ValidValues` _property_, and we use the property in Intellisense completers. Although `engineIntrinsics` contains current cmdlet info it does not contain current binding parameter info.
That way requires too much effort. 😕 Is there another way?

lzybkr commented:
We can add an interface:

```c#
public interface IExtendedValidateArguments
{
    void Validate(object arguments, EngineIntrinsics engineIntrinsics, /* extra args*/);
}
```

Then:
```C#
    public abstract class ValidateArgumentsAttribute : CmdletMetadataAttribute
    {
        //...

        internal void InternalValidate(object o, EngineIntrinsics engineIntrinsics, /* extra args */)
        {
            if (var extended = this as IExtendedValidateArguments) {
                extended.Validate(o, engineIntrinsics, /* extra args */ );
            } else {
                Validate(o, engineIntrinsics);
            }
        }
```

And lastly implement that interface for ValidateSetAttribute.

Alternatively, we can move the type test to the parameter binder to avoid passing extra arguments if that is expensive somehow.

iSazonov commented:
Thanks! I still don't see how fix the `ValidValues`  property for completers. It seems we need a new method in the interface?

daxian-dbw reviewed (commented):
> It seems we catch a race condition on slow OSX.
Wasn't there a CLR bug we were hitting on Mac related to defining classes?

I'm not aware of any CLR bug that affects PowerShell class except for the crossgen one that causes powershell to hang. We don't crossgen for CI builds, so that shouldn't affect it here.

iSazonov commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,19 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
+        /// Initializes a new instance of the CachedValidValuesGeneratorBase class.
+        /// </summary>
+        /// <param name="cacheExpirationInSeconds">Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.</param>
+        public CachedValidValuesGeneratorBase(int cacheExpirationInSeconds)
```
Fixed.

iSazonov commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1354,19 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
+        private IValidateSetValuesGenerator validValuesGenerator = null;
+
+        // The valid values generator cache works across 'ValidateSetAttribute' instances.
+        private static ConcurrentDictionary<Type, IValidateSetValuesGenerator> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, IValidateSetValuesGenerator>();
```
Above I tried to add cleanup but @lzybkr concluded that it didn't matter - we don't expect a huge number of generators.

daxian-dbw replied:
#Closed

iSazonov commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1380,13 +1446,28 @@ public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
+                    return _validValues;
+                }
+
+                var validValuesLocal = validValuesGenerator.GetValidValues();
```
It seems related #3769 - completers silently falls back to paths.
I tested a generator with `throw "Test throw!"` - parameter binder handles any exceptions:
```powershell
Get-TestValidateSetPS4 : Cannot validate argument on parameter 'Param1'. Test throw!
At line:1 char:32
+ Get-TestValidateSetPS4 -param1 fdfd
+                                ~~~~
    + CategoryInfo          : InvalidData: (:) [Get-TestValidateSetPS4], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Get-TestValidateSetPS4
```
If the generator throws good formatted exception users get it in InnerException.
So it seems redundant to add extra level to catch the exception.

daxian-dbw replied:
Agreed that wrapping the exception is not necessary.

iSazonov commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1533,44 @@ public ValidateSetAttribute(params string[] validValues)
+        /// <exception cref="ArgumentException">for null arguments</exception>
+        public ValidateSetAttribute(Type valuesGeneratorType)
+        {
+            if (!typeof(IValidateSetValuesGenerator).IsAssignableFrom(valuesGeneratorType))
```
It is useful for developers only - so maybe we skip the reflection?

daxian-dbw replied:
I think basic validations are necessary. For example, do you allow people to use an internal/private type as an argument? `Activator.CreateInstance` works on non-public types but I don't think we should support non-public types.

daxian-dbw replied:
And I think when validation fails, the thrown exception should be informational. `NewArgumentException("valuesGeneratorType")` doesn't indicate what is the problem.

iSazonov replied:
I added a check IsNonPublic.

We use the same pattern ` NewArgumentException` in all constructors in the file. And the exception is intended for developers only not users.

daxian-dbw replied:
#Closed

iSazonov commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1533,44 @@ public ValidateSetAttribute(params string[] validValues)
+
+            s_ValidValuesGeneratorCache.TryGetValue(valuesGeneratorType, out IValidateSetValuesGenerator ValidValuesGeneratorCacheEntry);
+
+            if (ValidValuesGeneratorCacheEntry == null)
```
Fixed.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1357,7 +1377,7 @@ internal static Attribute GetAttribute(AttributeAst attributeAst)
             {
                 throw InterpreterError.NewInterpreterException(attributeAst, typeof(RuntimeException), attributeAst.Extent,
-                    "CustomAttributeTypeNotFound", ParserStrings.CustomAttributeTypeNotFound, attributeAst.TypeName.FullName);
+                    "TypeNotFound", ParserStrings.TypeNotFound, attributeAst.TypeName.FullName);
```
It is @lzybkr request https://github.com/PowerShell/PowerShell/pull/3784#discussion_r124377253

daxian-dbw replied:
> I think I'd reuse TypeNotFound instead of adding a new message.

But `CustomAttributeTypeNotFound` is not a new resource string, and it looks to me fit pretty well in the `if (attributeType == null)` situation:
`Cannot find the type for custom attribute '{0}'. Make sure that the assembly that contains this type is loaded.`

iSazonov replied:
@lzybkr  @daxian-dbw Please make a final conclusion - I can't throw a coin. :-)

lzybkr replied:
My comment was for a different place in the code and I stand by that comment. This location should not be changed.

iSazonov replied:
Sorry, I lost the thread.

Reverted.

iSazonov committed:
Make constructor protected. Fix 'if'.

daxian-dbw commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1532,42 @@ public ValidateSetAttribute(params string[] validValues)
+                s_ValidValuesGeneratorCache.TryAdd(valuesGeneratorType, ValidValuesGeneratorCacheEntry);
+            }
+
+            validValuesGenerator = ValidValuesGeneratorCacheEntry;
```
How about replacing the whole `if {}` block with 
```
validValuesGenerator = s_ValidValuesGeneratorCache.GetOrAdd(valuesGeneratorType, (key) =>(IValidateSetValuesGenerator)Activator.CreateInstance(key))
```

iSazonov replied:
Good catch!

Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1455,6 +1533,44 @@ public ValidateSetAttribute(params string[] validValues)
+                // Add a valid values generator to the cache.
+                // We don't cache valid values.
+                // We expect that valid values can be cached in the valid values generator.
+                ValidValuesGeneratorCacheEntry = (IValidateSetValuesGenerator)Activator.CreateInstance(valuesGeneratorType);
```
#Close

iSazonov committed:
Add a check Type.IsNonPublic. Replace 'if' with GetOrAdd()

powercode commented:
After thinking a bit more about this, what I would want is `[ValidateCompleterSet]`

It would combine two features: `IArgumentCompleter` for getting values, and `ValidateSet` for limiting what can be passed to the function/cmdlet.

That way we could also get the tooltips etc, and have the validation based on the `CompletionText` of the `CompletionResult`.

I understand I'm a bit late to this party...

iSazonov commented:
@lzybkr I believe we should stop and merge because here we cannot resolve the problem with ValidValues public property while we share valid values generators.
It seems the @powercode's idea (thanks!) opens alternative way.

lzybkr commented:
I'm fine with merging soon - but the test failure needs to be worked out.

I'll be very busy for the next 1.5 weeks, much of the time offline, so don't be too concerned if you don't see any response from me.

daxian-dbw commented on test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1:
```diff
@@ -216,3 +216,263 @@ Describe 'Type resolution with attributes' -Tag "CI" {
+        It 'Can implement CachedValidValuesGeneratorBase with cache expiration in PowerShell' {
+            Get-TestValidateSetPS5 -Param1 "TestString1" -ErrorAction SilentlyContinue | Should BeExactly "TestString1"
+            Get-TestValidateSetPS5 -Param1 "TestString1" -ErrorAction SilentlyContinue | Should BeExactly "TestString1"
+            Start-Sleep 2
```
I think the failure in CI build is due to this. You can either change to `base(1)`, or make it sleep for 3 seconds (not sure which is more reliable).

iSazonov replied:
I am not happy with the solution but set 3 seconds.

daxian-dbw commented:
I think this is almost ready to be merged -- 2 more comments to be addressed.
> I believe we should stop and merge because here we cannot resolve the problem with ValidValues public property while we share valid values generators.

I don't have the context of the `'the problem with ValidValues'`, can you please elaborate it a bit? But no matter what, it should not block this PR and the further discussion should be captured in an issue.

iSazonov commented:
@daxian-dbw
From lzybkr optional idea:
> I could imagine wanting to know the command, maybe the command ast, the actual parameter the attribute is applied to, and maybe the already bound parameters.
...
We can add an interface:
public interface IExtendedValidateArguments

The idea is to pass a context to a valid value generator. 
Using the new interface we can pass the context to `Validate()` but we have public property ValidValues - I don`t see how we can pass the context to the property. The root of the problem is that we cache and share the generator. If we add "context" to the cache, the effectiveness of the cache will decrease to 0.

iSazonov committed:
Revert exception.

iSazonov committed:
Fix test

iSazonov commented:
/cc @joeyaiello  We are near a Milstone - 996 repo forks :-)

daxian-dbw reviewed (approved):
LGTM.

daxian-dbw commented:
I will merge this PR soon. I assume #3744 is fixed by this PR, so for further enhancement discussion, please open a new issue to continue.

@iSazonov now that the design and implementation are complete, could you please update the PR description with a summary of the design and new functionalities? We need that information for the documentation.

iSazonov commented:
@daxian-dbw I added the PR description - please review.

daxian-dbw commented:
@iSazonov Perfect! Thank you for the great work!

iSazonov commented:
Many thanks @lzybkr for ideas and helps!

Menelion commented:
Hi everyone,  
I found this PR when searching for exactly the same solution: how to implement `validateSet` from a dynamically generated array of subfolders in a PS script. I see that this is merged, but is there any documentation about this? If not, could you please either point me to the test where I can see it in action or just drop a couple lines to illustrate it? And one more question: if I run Windows 10 1709, do I need to update PowerShell to have this working?  
Thanks!

markekraus commented:
@Menelion Generally you should open a new issue for this kind of thing rather than comment on a closed PR you can reference a PR in an issue by doing `#3784` and it will appear like this: #3784 

You can find some documentation of the API here https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.ivalidatesetvaluesgenerator?view=pscore-6.0.0

This features was added in PowerShell Core 6.0.0. To use, your PowerShell code must run in PowerShell Core 6.0.0 or newer. 

You can find example usage here in this PR or at https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1

iSazonov commented:
>  where I can see it in action or just drop a couple lines to illustrate it?

@Menelion You could see our tests which was added in the PR.

felixfbecker commented:
I am trying to use this, but I always get `The type or namespace name 'IValidateSetValuesGenerator' could not be found (are you missing a using directive or an assembly reference?)`

I am using the latest `PowerShellStandard.Library` (`5.1.0-preview-06`). How can I reference this interface?

lzybkr commented:
That interface was introduced in `6.0.0`, so you would need to reference to https://powershell.myget.org/feed/powershell-core/package/nuget/System.Management.Automation
